cache-invalidation/duffy.py

```diff
@@ -0,0 +1,164 @@
+server_env+=" GERRIT_PROJECT='%s'" % os.getenv("GERRIT_PROJECT")
+server_env+=" GERRIT_REFSPEC='%s'" % os.getenv("GERRIT_REFSPEC")
+server_env+=" YUM_REPO='%s'" % os.getenv("YUM_REPO", "")
+server_env+=" GLUSTER_VOLUME='%s'" % os.getenv("EXPORT")
+server_env+=" ENABLE_ACL='%s'" % os.getenv("ENABLE_ACL", "")
+
+# add the export with environment to ~/.bashrc for server-1
+cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+tee -a ~/.bashrc' <<< "%s"
+""" % (b['hosts'][0], server_env)
+subprocess.call(cmd, shell=True)
+
+cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+	yum -y install curl &&
+	curl -o server1 %s && bash server1 0
+'""" % (b['hosts'][0], server_script)
+rtn_code=subprocess.call(cmd, shell=True)
+
+# add the export with environment to ~/.bashrc for server-2
+if rtn_code == 0:
+    cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+    tee -a ~/.bashrc' <<< "%s"
+    """ % (b['hosts'][1], server_env)
+    subprocess.call(cmd, shell=True)
+
+    cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+    	    yum -y install curl &&
+	    curl -o server2 %s && bash server2 0 0 0 %s
+    '""" % (b['hosts'][1], server_script, b['hosts'][0])
+    rtn_code=subprocess.call(cmd, shell=True)
+
+# check rtn_code and skip resume client-1 part after failure
+if rtn_code == 0:
+    cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+	    yum -y install curl &&
+	    curl -o server1 %s && bash server1 - - 0
+    '""" % (b['hosts'][0], server_script)
+    rtn_code=subprocess.call(cmd, shell=True)
+
+# check rtn_code and skip client environment part after failure
+if rtn_code == 0:
+    # NFS-Client (parameters need double escape, passed on ssh commandline)
+    client_env="export SERVER='%s'" % b['hosts'][0]
+    client_env+=" EXPORT='/%s'" % os.getenv("EXPORT")
+    client_env+=" TEST_PARAMETERS='%s'" % os.getenv("TEST_PARAMETERS", "")
+
+    # add the export with environment to ~/.bashrc for client-1
+    cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+        tee -a ~/.bashrc' <<< "%s"
+        """ % (b['hosts'][2], client_env)
+    rtn_code=subprocess.call(cmd, shell=True)
+
+    # NFS-Client (parameters need double escape, passed on ssh commandline)
+    client_env="export SERVER='%s'" % b['hosts'][1]
+    client_env+=" EXPORT='/%s'" % os.getenv("EXPORT")
+    client_env+=" TEST_PARAMETERS='%s'" % os.getenv("TEST_PARAMETERS", "")
+
+    # add the export with environment to ~/.bashrc for client-2
+    cmd="""ssh -t -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@%s '
+        tee -a ~/.bashrc' <<< "%s"
+        """ % (b['hosts'][3], client_env)
+    rtn_code=subprocess.call(cmd, shell=True)
+
+ # check rtn_code and skip rest test part after failure
+if rtn_code == 0:
+    versions = [3 , 4.0, 4.1]
```

Comment:
Ah, I guess the major change is here. Most of these should probably be part of the same client.sh script, which can run the tests in sequence and abort in case something fails.

The intention of this script should be to start something on the client that does the real testing. Merging test cases with the setup and test-script seem not so optimal and make maintaining these tests more complex. Please try to make this a little simpler.

You could suggest passing a space separated list of `NFS_VERSIONS` environment variable from the calling test-case to this script and forward that into the `client_env` to the script. Or have the script called once per value in `NFS_VERSIONS`.